File: bend2/effs/chan_recv.js
// Chan
// ====

function chan_recv(handle, k) {
  const row = io_read(handle, "chan");
  if (row === null) {
    return { $: "None" };
  }
  if (row.ring.length > 0) {
    const v = chan_take(row);
    if (row.shut && row.ring.length === 0) {
      io_kill(handle);
    }
    return { $: "Some", value: v };
  }
  if (row.wait.length > 0 && row.wait[0].item !== null) {
    return { $: "Some", value: chan_wake(row, true) };
  }
  if (row.shut) {
    io_kill(handle);
    return { $: "None" };
  }
  row.wait.push({ cont: k, item: null });
  return;
}

File: bend2/effs/chan_recv.c
// Chan
// ====
//! use ./chan.c

Term chan_recv_run(Env e, Term* f, IoWork* w) {
  IoHand   h   = io_hand_c(e, f[0]);
  ChanRow* row = chan_at(h);
  if (row == NULL) {
    return term_pak(CID_NONE, 0);
  }
  if (row->size > 0) {
    Term v = chan_take(row);
    if (row->shut && row->size == 0) {
      chan_free(h, row);
    }
    return chan_some(e, v);
  }
  if (row->wait != NULL && row->wait->item != TERM_HOLE) {
    return chan_some(e, chan_wake(row, chan_bool(true)));
  }
  if (row->shut) {
    chan_free(h, row);
    return term_pak(CID_NONE, 0);
  }
  chan_park(row, f[1], TERM_HOLE);
  return IO_PARK;
}

static void __attribute__((constructor)) chan_recv_use(void) {
  io_eff(FID_CHAN_RECV, CID_CHAN_RECV, chan_recv_run, 0);
}
